x/stakeibc/keeper/icacallbacks_redemption_test.go

```diff
@@ -0,0 +1,212 @@
+package keeper_test
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	_ "github.com/stretchr/testify/suite"
+
+	recordtypes "github.com/Stride-Labs/stride/x/records/types"
+	stakeibckeeper "github.com/Stride-Labs/stride/x/stakeibc/keeper"
+	"github.com/Stride-Labs/stride/x/stakeibc/types"
+	stakeibc "github.com/Stride-Labs/stride/x/stakeibc/types"
+)
+
+type RedemptionCallbackState struct {
+	epochUnbondingNumbers   []uint64
+	userRedemptionRecordIds []string
+	epochNumber             uint64
+}
+
+type RedemptionCallbackArgs struct {
+	packet channeltypes.Packet
+	ack    *channeltypes.Acknowledgement
+	args   []byte
+}
+
+type RedemptionCallbackTestCase struct {
+	initialState RedemptionCallbackState
+	validArgs    RedemptionCallbackArgs
+}
+
+func (s *KeeperTestSuite) SetupRedemptionCallback() RedemptionCallbackTestCase {
+	epochNumber := uint64(1)
+
+	// userRedemptionRecords should NOT be claimable until after the callback is called
+	recordId1 := recordtypes.UserRedemptionRecordKeyFormatter(chainId, epochNumber, "sender")
+	userRedemptionRecord1 := recordtypes.UserRedemptionRecord{
+		Id:          recordId1,
+		IsClaimable: false,
+	}
+	recordId2 := recordtypes.UserRedemptionRecordKeyFormatter(chainId, epochNumber, "other_sender")
+	userRedemptionRecord2 := recordtypes.UserRedemptionRecord{
+		Id:          recordId2,
+		IsClaimable: false,
+	}
+
+	// the hostZoneUnbonding should have HostZoneUnbonding_UNBONDED - meaning unbonding has completed, but the tokens
+	// have not yet been transferred to the redemption account
+	hostZoneUnbonding := recordtypes.HostZoneUnbonding{
+		HostZoneId:            chainId,
+		Status:                recordtypes.HostZoneUnbonding_UNBONDED,
+		UserRedemptionRecords: []string{recordId1, recordId2},
+	}
+
+	epochUnbondingRecord := recordtypes.EpochUnbondingRecord{
+		EpochNumber:        epochNumber,
+		HostZoneUnbondings: []*recordtypes.HostZoneUnbonding{&hostZoneUnbonding},
+	}
+	hostZone := stakeibc.HostZone{
+		ChainId:        chainId,
+		HostDenom:      atom,
+		IBCDenom:       ibcAtom,
+		RedemptionRate: sdk.NewDec(1.0),
+	}
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx(), hostZone)
+	s.App.RecordsKeeper.SetEpochUnbondingRecord(s.Ctx(), epochUnbondingRecord)
+	s.App.RecordsKeeper.SetUserRedemptionRecord(s.Ctx(), userRedemptionRecord1)
+	s.App.RecordsKeeper.SetUserRedemptionRecord(s.Ctx(), userRedemptionRecord2)
+
+	packet := channeltypes.Packet{}
+	var msgs []sdk.Msg
+	msgs = append(msgs, &banktypes.MsgSend{})
+	ack := s.ICAPacketAcknowledgement(msgs)
+	callbackArgs := types.RedemptionCallback{
+		HostZoneId:            chainId,
+		UnbondingEpochNumbers: []uint64{epochNumber},
```

Comment:
nit: should we update this proto field to `EpochUnbondingRecordIds` to align with undelegate?

Reply:
done